core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java

```diff
@@ -96,20 +104,34 @@ private void calculateData(IItemCollection input) {
 			IAggregator<?, ?> aggregator = Aggregators.distinct(query.getGroupBy());
 			Set<?> aggregate = input.getAggregate((IAggregator<Set<?>, ?>) aggregator);
 			if (aggregate != null) {
-				for (Object o : aggregate) {
-					IItemCollection rowCollection = input.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
-					Object[] row = newRow();
-					int column = 0;
-					for (; column < attributes.size(); column++) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				int SUM_NUM_THREADS = Runtime.getRuntime().availableProcessors();
+				ExecutorService exec = Executors.newFixedThreadPool(SUM_NUM_THREADS);
+				try {
+					for (final Object o : aggregate) {
+						exec.submit(new Runnable() {
+							@Override
+							public void run() {
+								IItemCollection rowCollection = newInput
+										.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
+								Object[] row = newRow();
+								int column = 0;
+								for (; column < attributes.size(); column++) {
+									// Optimization - it is too expensive to do aggregation for these. You simply
+									// get first non-null matching attribute - we're only using this for the group by today.
+									row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+								}
+								for (int j = 0; j < aggregators.size(); j++) {
+									row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+								}
+								data.add(row);
+							}
+						});
 					}
-					for (int j = 0; j < aggregators.size(); j++) {
-						row[column + j] = rowCollection.getAggregate(aggregators.get(j));
-					}
-					data.add(row);
+				} finally {
+					exec.shutdown();
 				}
+				// Higher timeout value added for worst case
+				exec.awaitTermination(1, TimeUnit.HOURS);
```

Comment:
It may be worth to make this value configurable. IE let hte use know that how long it will allow for the background job to run.

Also, it may be worthy to actually enforce the shutdown once this delay is over. IE soemthing like that

```java
try { 
  // allow for budget (default 1h) to process the aggregates
  if(!exec.awaitTermination(delay, delayUnit)) {
    exec.shutdownNow();
    // last grace period
    if (!pool.awaitTermination(1, TimeUnit.MINUTES)) {
      // report an issue with the pool
    }
  }
} catch (InterruptedException e) {
  Thread.currentThread().interrupt();
}
```

Also I think this code could moved in the same scope as the shutdown call.

Reply:
Agreed, let's add a preference. Also, the comment is not aligned with the implementation, seems it's set to 1 minute, not 1 hour.

Reply:
@thegreystone Removed the hard-coding and added it as part of preferences. Please approve the changes if you feel its good to go now.